### PR TITLE
Fixes and Updates for C++ Command Line Module

### PR DESCRIPTION
**Pull request by mizlan**
More aligned with the spec and some pointers and tips. Also removed some potentially dangerous advice and other unorthodox tips in favor of best practices.

### CODE REVIEW

File: content/1_General/Cpp_Command.mdx [hunk 1]
```diff
@@ -159,29 +159,32 @@ If you want to code in (neo)vim, you can install WSL and code through WSL bash.
...
+    ```bash
+    alias g++=g++-10
     ```
-    cd /usr/local/bin
-    ln -fs g++-9 g++
-    ```
 
-    `g++ --version` should now output the same thing as `g++-9 --version`.
+    Once you do so, `g++ --version` should now output the same thing as `g++-9 --version`.
+
```
Comment (thecodingwizard) [inline content/1_General/Cpp_Command.mdx:184]: should this be `g++-10 --version`?

File: content/1_General/Cpp_Command.mdx [hunk 2]
```diff
@@ -286,32 +302,32 @@ Of course, retyping the flags above can get tedious. You should define shortcuts
...
+subl ~/.zshrc
 ```
 
 You can add **aliases** and **functions** here, such as the following to compile and run C++ on Mac.
 
 ```
-co() { g++ -std=c++17 -O2 -o $1 $1.cpp -Wall; }
-run() { co $1 && ./$1 & fg; }
+co() { g++ -std=c++17 -O2 -o "${1%.*}" $1 -Wall; }
+run() { co $1 && ./${1%.*} & fg; }
```
Comment (thecodingwizard) [inline content/1_General/Cpp_Command.mdx:327]: Isn't the original shell script preferable since it safes you a few keystrokes (you don't have to type .cpp)?

Or is there something I'm missing? (oops I don't use shell scripts)

**Comment by thecodingwizard** [comment]
Thanks for the PR! 😃

File: content/1_General/Cpp_Command.mdx [hunk 3]
```diff
@@ -136,7 +136,7 @@ If you want to code in (neo)vim, you can install WSL and code through WSL bash.
 
     If you previously installed these you may need to update them:
 
-    ```
+    ```bash
```
Comment (thecodingwizard) [inline content/1_General/Cpp_Command.mdx:139]: I don't believe we have syntax highlighting for bash (I think). (Keep the change though; we might add it in the future.)

If there is sufficient need for bash syntax highlighting we can add it, though this will increase the bundle size.

File: content/1_General/Cpp_Command.mdx (hunk 2 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:327]: As hinted at in the commit message, this makes completion easier (tab completion normally works by files). The original method would imply completing and then deleting the `.cpp` from the end if that makes sense.

File: content/1_General/Cpp_Command.mdx (hunk 3 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:139]: Makes sense; I would leave it out. The benefit is very minimal.

File: content/1_General/Cpp_Command.mdx (hunk 1 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:184]: Yup, thanks for catching that! Will run through another time to make sure everything got updated.

**Comment by mizlan** [comment]
This can be merged at your discretion; I don't have any other changes planned.

File: content/1_General/Cpp_Command.mdx [hunk 4]
```diff
@@ -159,29 +159,32 @@ If you want to code in (neo)vim, you can install WSL and code through WSL bash.
...
-    cd /usr/local/bin
-    ln -fs g++-9 g++
+    ```bash
+    alias g++=g++-10
     ```
 
-    `g++ --version` should now output the same thing as `g++-9 --version`.
+    Once you do so, `g++ --version` should now output the same thing as `g++-10 --version`.
+
+    Note: avoid overriding the system `g++` with symlinking or hard-linking as
```
Comment (bqi343) [inline content/1_General/Cpp_Command.mdx:185]: >  Note: avoid overriding the system `g++` with symlinking or hard-linking as
    that will almost surely cause problems. Don't worry if you don't know what
    those terms mean.

what problems?

File: content/1_General/Cpp_Command.mdx [hunk 5]
```diff
@@ -234,9 +237,23 @@ If you type some integer and then press enter, then the program should produce o
 g++ name.cpp -o name && ./name
 ```
 
+Don't be afraid, the `./name` will only run if `g++ name.cpp -o name` finished successfully. Such is the magic of `&&`!
```
Comment (bqi343) [inline content/1_General/Cpp_Command.mdx:240]: > Don't be afraid,

??

File: content/1_General/Cpp_Command.mdx (hunk 4 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:185]: System build processes which expect clang.

File: content/1_General/Cpp_Command.mdx (hunk 5 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:240]: That is to say that you need not worry about `./name` running if the compilation wasn't successful.

File: content/1_General/Cpp_Command.mdx (hunk 4 above)
Comment (bqi343) [reply content/1_General/Cpp_Command.mdx:185]: Could you give an example?

(oops why don't they just use `clang++` then)

**Comment by bqi343** [comment]
looks like the prerequisite module should be updated to `running-code-locally` instead of `running-code`

File: content/1_General/Cpp_Command.mdx (hunk 4 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:185]: It's a best practice to not override system packages in general, not limited to just `g++`.

Comment (bqi343) [reply content/1_General/Cpp_Command.mdx:185]: still waiting on a concrete example

why not override `g++` with the *real* `g++`? :/

File: content/1_General/Cpp_Command.mdx (hunk 5 above)
Comment (thecodingwizard) [reply content/1_General/Cpp_Command.mdx:240]: Maybe rephrase it into something slightly more formal to flow better with the rest of the module? (Maybe just say `&&` only executes the second command if the first command is successful, so `./name` will only run if the compilation finished successfully. or something?)

File: content/1_General/Cpp_Command.mdx (hunk 4 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:185]: For the same reason Apple chose to link `g++` to `clang++`—probably not a very good reason, but we will never know. :/ Also I just realized it was `/usr/local/bin` versus `/usr/bin`, but the issue of `PATH` resolution still exists. As such, it will always be safer to only activate this in _interactive_ shells, which is what aliasing in the rc file does. In addition, the creation of such a link in `/usr/local/bin` is now not managed by Homebrew, which may lead to a dangling symlink if the target executable is (re)moved.

As for a concrete example, `g++` is not really seen as much in the wild anymore. So I guess you could say this is overblown, but the fact is it's not best practice for multiple reasons.

Comment (bqi343) [reply content/1_General/Cpp_Command.mdx:185]: tbh I don't completely understand all of this but seems alright :)

File: content/1_General/Cpp_Command.mdx [hunk 6]
```diff
@@ -254,11 +271,11 @@ g++ -std=c++17 -O2 name.cpp -o name -Wall
 
 Explanation:
 
-- `-O2` tells `g++` to compile your code to run more quickly (see [here](https://www.rapidtables.com/code/linux/gcc/gcc-o.html))
+- `-O2` tells `g++` to compile your code to run more quickly (see [here](https://www.rapidtables.com/code/linux/gcc/gcc-o.html)). Note that `-O2` will slow down compilation time. Think of it as a trade-off between runtime and compile time. 
 - `-std=c++17` allows you to use features that were added to C++ in 2017. USACO currently uses `-std=c++11`.
 - `-Wall` checks your program for common errors. See [Debugging](/general/debugging-lang/#warning-options) for more information.
 
-You should always compile with these flags.
+You should always compile with these flags (with the possible exception of `-O2`, as described above).
```
Comment (bqi343) [inline content/1_General/Cpp_Command.mdx:278]: Is the increase in compile time so significant?

File: content/1_General/Cpp_Command.mdx (hunk 5 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx]: Absolutely, will change it.

File: content/1_General/Cpp_Command.mdx (hunk 6 above)
Comment (mizlan) [reply content/1_General/Cpp_Command.mdx:278]: Probably not horribly significant, revert that if you want.
